keps/sig-cloud-provider/20200928-extend-konnectivity-for-both-directions.md

```diff
@@ -0,0 +1,194 @@
+* **Node Network** An IP reachable network space which contains all the clusters Nodes, for alpha. Worth noting that the Node Network may be completely disjoint from the Master network. It may have overlapping IP addresses to the Master Network or other means of network isolation. Direct IP routability between cluster and master networks should not be assumed. Later versions may relax the all node requirement to some.
+* **KAS** Kubernetes API Server, responsible for serving the Kubernetes API to clients.
+* **Konnectivity Server** The proxy server which runs in the master network. It has a secure channel established to the cluster network. It could work on either a HTTP Connect mechanism or gRPC. If the former it would expose a gRPC interface to KAS to provide connectivity service. If the latter it would use standard HTTP Connect. Formerly known as the Network Proxy Server.
+* **Konnectivity Agent** A proxy agent which runs in the node network for establishing the tunnel. Formerly known as the Network Proxy Agent.
+* **Flat Network** A network where traffic can be successfully routed using IP. Implies no overlapping (i.e. shared) IPs on the network.
+
+## Proposal
+
+Currently the Konnectivity Server is accepting requests from the KAS either with the gRPC or the HTTP Connect interfaces and is taking care of forwarding the traffic to the Konnectivity Agent using the previously established connections (initiated by the Agents). 
+
+In order to enable traffic from Kubelets and Pods running on Master Network, the Konnectivity Agents have to expose an endpoint that will be listening on a specific port for each of the destinations on the Master Network. As opposed to the traffic flowing from the Master Network to the Node Network, the Konnectivity Agent should act transparently: From a Kubelets or Pods standpoint, the Konnectivity Agent should be the final destination instead of acting as a proxy. 
+
+The reason why we do not intend to use the same strategy used in the other direction is that we do not have control over the clients using the Kubernetes default service to communicate with the KAS.
+
+### Traffic Flow
+
+```
+client =TCP=> (:6443) agent GRPC=> server =TCP=> KAS(:6443)
+                         |            ^
+                         |   Tunnel   |
+                         +------------+
+```
+
+The agent listens for TCP connections at a specific port for each configured destination. When a connection request is received by the Konnectivity Agent the following happens:
+1. A GRPC DIAL_REQ message is sent to the Konnectivity server containing the destination address associated with the current port.
+2. Upon reception of the DIAL_REQ the Konnectivity Server opens a TCP connection with the destination host/port and replies to the Konnectivity Agent with a GRPC DIAL_RES message.
+3. At this point the tunnel is established and data is piped through it, carried over GRPC DATA packets.
```

Comment:
I guess this has big performance impact. Because the other end is always `KAS:6443`, It seems a reverse proxy on konnectivity server side could do the same job without agent.

Reply:
Indeed, we have the same performance penalty that we are paying with traffic going from the Konnectivity server to the nodes. 
When using the reverse proxy, we don't have the additional layer of security provided by the authentication between the Konnectivity Agent and the Konnectivity Server but also we lose the ability to use SNI for load balancing.

Reply:
I guess a whitelist is necessary this sort of proxy in reserve order. master network has access everything in node network, but I am not sure the reverse is true.

Reply:
+1 to having an explicit allow list on the Konnectivity Server which controls where it will allow traffic to be sent on the control plane.

Reply:
@anfernee @cheftako 
That is correct and we have taken this into account in the [Risks and Mitigations / Allow list](https://github.com/kubernetes/enhancements/pull/2025/files#diff-5994260c1d3432760f8f28373c7cffc9cd5f9f935615ad6875071c7264361c38R145-R168) section.